packages/plugins/src/hardhat/simulate.ts

```diff
@@ -482,26 +482,12 @@ export const handleSimulationSuccess = async (
 export const createHardhatEthersProviderProxy = (
   ethersProvider: HardhatEthersProvider
 ): HardhatEthersProvider => {
-  return new Proxy(ethersProvider, {
+  const proxy = new Proxy(ethersProvider, {
     get: (target, prop) => {
       return (...args: any[]) => {
-        // Queue a snapshot of the Hardhat node state. We may revert to this snapshot later in this
-        // function to prevent a 'nonce too low' bug in Hardhat. We must queue the snapshot before
-        // calling `target[prop](...args)` to avoid this nonce bug, which occurs in spite of the
-        // fact that we don't immediately await either of these calls. We don't await this call yet
-        // because the `target[prop](...args)` call may be synchronous. If we await this call, then
-        // every call routed through this Proxy will become asynchronous, which could lead to
-        // expected behavior.
-        //
-        // More info on the nonce error is in this pull request description:
-        // https://github.com/sphinx-labs/sphinx/pull/1565.
-        const snapshotIdPromise = target.send('evm_snapshot', [])
-
-        // Call the method on the EthersJS provider.
-        const initialResult = target[prop](...args)
-        // Check if the call is synchronous. If so, return it.
-        if (!(initialResult instanceof Promise)) {
-          return initialResult
+        // Return the result directly if the method isn't asynchronous.
+        if (!sphinxCoreUtils.isPublicAsyncMethod(ethersProvider, prop)) {
```

Comment:
We use this to check whether the property is an asynchronous method. Our strategy of queueing the `initialResult` and `snapshotIdPromise` didn't work because of how Hardhat handles promises. Specifically, this situation led to a rate limit:
1. We queue the initial snapshot ID promise. Note that we don't use await: `const snapshotIdPromise = target.send('evm_snapshot', [])`
2. We queue the proxy call promise. We also don't use await: `target[prop](...args)`
3. We await the snapshot ID promise: `await snapshotIdPromise`. Crucially, this `await` caused the promise from step 2 to be awaited too. This made it possible to hit a rate limit here.

I fixed this by removing the logic that queues these promises in favor of using `isPublicAsyncMethod`, which simplifies this logic